openmc/lib/weight_windows.py

```diff
@@ -0,0 +1,311 @@
+from collections.abc import Mapping
+from ctypes import c_double, c_int, c_int32, c_char_p, c_size_t, POINTER
+from weakref import WeakValueDictionary
+
+import numpy as np
+from numpy.ctypeslib import as_array
+
+from openmc.exceptions import AllocationError, InvalidIDError
+from . import _dll
+from .core import _FortranObjectWithID
+from .error import _error_handler
+from .filter import EnergyFilter, MeshFilter, ParticleFilter
+from .mesh import _get_mesh
+from .mesh import meshes
+from .particle import ParticleType
+
+
+__all__ = ['WeightWindows', 'weight_windows_map']
+
+_dll.openmc_extend_weight_windows.argtypes = [c_int32, POINTER(c_int32), POINTER(c_int32)]
+
+_dll.openmc_update_weight_windows_magic.argtypes = 2*[c_int32] + [c_char_p] + 2*[c_double]
+_dll.openmc_update_weight_windows_magic.restype = c_int
+_dll.openmc_update_weight_windows_magic.errcheck = _error_handler
+
+_dll.openmc_weight_windows_size.restype = c_size_t
+
+_dll.openmc_weight_windows_get_index.argtypes = [c_int32, POINTER(c_int32)]
+_dll.openmc_weight_windows_get_index.restype = c_int
+_dll.openmc_weight_windows_get_index.errcheck = _error_handler
+
+_dll.openmc_weight_windows_get_id.argtypes = [c_int32, POINTER(c_int32)]
+_dll.openmc_weight_windows_get_id.restype = c_int
+_dll.openmc_weight_windows_get_id.errcheck = _error_handler
+
+_dll.openmc_weight_windows_set_id.argtypes = [c_int32, c_int32]
+_dll.openmc_weight_windows_set_id.restype = c_int
+_dll.openmc_weight_windows_set_id.errcheck = _error_handler
+
+_dll.openmc_weight_windows_set_mesh.argtypes = [c_int32, c_int32]
+_dll.openmc_weight_windows_set_mesh.restype = c_int
+_dll.openmc_weight_windows_set_mesh.errcheck = _error_handler
+
+_dll.openmc_weight_windows_get_mesh.argtypes = [c_int32, POINTER(c_int32)]
+_dll.openmc_weight_windows_get_mesh.restype = c_int
+_dll.openmc_weight_windows_get_mesh.errcheck = _error_handler
+
+_dll.openmc_weight_windows_set_energy_bounds.argtypes = [c_int32, POINTER(c_double), c_size_t]
+_dll.openmc_weight_windows_set_energy_bounds.restype = c_int
+_dll.openmc_weight_windows_set_energy_bounds.errcheck = _error_handler
+
+_dll.openmc_weight_windows_get_energy_bounds.argtypes = [c_int32, POINTER(POINTER(c_double)), POINTER(c_size_t)]
+_dll.openmc_weight_windows_get_energy_bounds.restype = c_int
+_dll.openmc_weight_windows_get_energy_bounds.errcheck = _error_handler
+
+_dll.openmc_weight_windows_set_particle.argtypes = [c_int32, c_char_p]
+_dll.openmc_weight_windows_set_particle.restype = c_int
+_dll.openmc_weight_windows_set_particle.errcheck = _error_handler
+
+_dll.openmc_weight_windows_get_particle.argtypes = [c_int32, POINTER(c_int)]
+_dll.openmc_weight_windows_get_particle.restype = c_int
+_dll.openmc_weight_windows_get_particle.errcheck = _error_handler
```

Comment:
Noting that I should cover this in testing...


Also, I updated `openmc_weight_windows_set_particle` to use an `int` instead of `char*` to pass the particle type. I think this reduces the number of hoops to jump through when using the CAPI, but let me know if you feel differently.

Reply:
This should now be covered in `test_ww_attrs`.